pkg/backup/remap_crd_version_action.go

```diff
@@ -0,0 +1,97 @@
+	logger logrus.FieldLogger
+}
+
+func NewRemapCRDVersionAction(logger logrus.FieldLogger) *RemapCRDVersionAction {
+	return &RemapCRDVersionAction{logger: logger}
+}
+
+func (a *RemapCRDVersionAction) AppliesTo() (velero.ResourceSelector, error) {
+	return velero.ResourceSelector{
+		IncludedResources: []string{"customresourcedefinition.apiextensions.k8s.io"},
+	}, nil
+}
+
+// TODO, if this works
+func (a *RemapCRDVersionAction) Execute(item runtime.Unstructured, backup *v1.Backup) (runtime.Unstructured, []velero.ResourceIdentifier, error) {
+	a.logger.Info("Executing RemapCRDVersionAction")
+
+	var crd apiextv1.CustomResourceDefinition
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(item.UnstructuredContent(), &crd); err != nil {
+		return nil, nil, errors.Wrap(err, "unable to convert unstructured item to CRD")
+	}
+
+	log := a.logger.WithField("plugin", "RemapCRDVersionAction").WithField("CRD", crd.Name)
+
+	// TODO: Is it possible or likely that a CRD had multiple versions without a Schema?
+	if len(crd.Spec.Versions) == 1 && crd.Spec.Versions[0].Schema == nil {
+		log.Debug("CRD is a candidate for v1beta1 backup")
+
+		// Two solutions are presented here, and they both appear to work. I've uncommented them both for readability; they won't necessarily compile this way.
+
+		// Solution 1: create empty structs where they're required, so that the definition is technically conformant.
+		// Pros:
+		//   - Compatible with the existing restore plugin for x-preserve-unknown-fields, which is necessary on v1 without a schema
+		//   - "upgrading" the CRD to v1 keeps the backup valid for longer, across more versions of Kubernetes
+		// Cons:
+		//   - Even though the structs are empty, they're not taking the data as it was.
+		crd.Spec.Versions[0].Schema = new(apiextv1.CustomResourceValidation)
+		crd.Spec.Versions[0].Schema.OpenAPIV3Schema = new(apiextv1.JSONSchemaProps)
+
+		newItem, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&crd)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "unable to convert crd to unstructured item")
+		}
+
+		return &unstructured.Unstructured{Object: newItem}, nil, nil
+	}
+
+	// Solution 2: Change the API version that's retained in the backup so that on restore, it will be restored as v1beta1.
+	// Pros:
+	//   - Relatively simple; just changing a string
+	//   - Data is the same as what's in the source Kubernetes cluster - the only reason we got it as a v1 instance was due to our usage of the preferred version endpoint.
+	// Cons:
+	//   - Data in the backup won't be valid for as long - v1beta1 CRDs will have to be upgraded sooner rather than later - but is that Velero's problem?
+
+	// Since we can't manipulate an Unstructured's Object map directly, get a copy to manipulate before setting it back
+	tempMap := item.UnstructuredContent()
+
+	if err := unstructured.SetNestedField(tempMap, "apiextensions.k8s.io/v1beta1", "apiVersion"); err != nil {
```

Comment:
does this actually work (namely, on the restore side)? I woulda thought there would be more logic required to convert a v1 CRD back to a v1beta1 that would actually be accepted by the API server on restore.

Reply:
So far in my tests, yes, both of these solutions will be accepted on restore.

Changing the version field appears to get it accepted and then returned via the `v1` API endpoint in much the same way as creating it via Helm does.

Reply:
Huh, interesting. I woulda thought there'd be incompatible fields in the v1 API that couldn't be submitted as v1beta1, but I guess this is what's meant by forwards- and backwards-compatible in Kubernetes API changes.

If this is valid for all scenarios, then I agree with @jenting that this is probably the preferable path

Reply:
> I woulda thought there'd be incompatible fields in the v1 API that couldn't be submitted as v1beta1, but I guess this is what's meant by forwards- and backwards-compatible in Kubernetes API changes.

Well, the problem was that these `v1beta1` CRDs were being returned from the v1 endpoint without a schema entry at all, then velero couldn't restore them. So the Kube API was happily serving them (even as they were invalid), but not accepting them back at the v1 endpoint.

Reply:
right, I get that - I'm just saying that behind the scenes, Kube has conversion logic to convert them from v1beta1 to v1 and vice-versa, so that they can be served from either API version. I'm surprised that you can just take the v1 YAML representation, change the version, and apply it directly to the v1beta1 version without going through the Kube conversion logic.

Reply:
Here's the diff of what I applied vs what I got back - the only things that got applied appear to be a `versions` slice and the `PerserveUnknownFields` field. All of these things are valid in v1beta1.

`diff authcodes.yaml from_v1.yaml`

```diff
1c1
< apiVersion: apiextensions.k8s.io/v1beta1
---
> apiVersion: apiextensions.k8s.io/v1
3a4,8
>   annotations:
>     kubectl.kubernetes.io/last-applied-configuration: |
>       {"apiVersion":"apiextensions.k8s.io/v1beta1","kind":"CustomResourceDefinition","metadata":{"annotations":{},"name":"authcodes.dex.coreos.com"},"spec":{"group":"dex.coreos.com","names":{"kind":"AuthCode","listKind":"AuthCodeList","plural":"authcodes","singular":"authcode"},"version":"v1"}}
>   creationTimestamp: "2020-02-12T22:13:38Z"
>   generation: 1
4a10,12
>   resourceVersion: "3645"
>   selfLink: /apis/apiextensions.k8s.io/v1/customresourcedefinitions/authcodes.dex.coreos.com
>   uid: f73f31e4-4613-4f45-9498-adff20be05a1
5a14,15
>   conversion:
>     strategy: None
12c22,46
<   version: v1
---
>   preserveUnknownFields: true
>   scope: Namespaced
>   versions:
>   - name: v1
>     served: true
>     storage: true
> status:
>   acceptedNames:
>     kind: AuthCode
>     listKind: AuthCodeList
>     plural: authcodes
>     singular: authcode
>   conditions:
>   - lastTransitionTime: "2020-02-12T22:13:38Z"
>     message: no conflicts found
>     reason: NoConflicts
>     status: "True"
>     type: NamesAccepted
>   - lastTransitionTime: "2020-02-12T22:13:38Z"
>     message: the initial names have been accepted
>     reason: InitialNamesAccepted
>     status: "True"
>     type: Established
>   storedVersions:
>   - v1
```